Trap2 time stepper

valeriabarra opened the pull request:
### Description

A HUGE thank you goes to @thomasgibson, who helped enormously in this PR. 

This PR adds another time integrator to our family of time stepping schemes. This new time integrator `Trap2LockWoodWeller` is described in the reference [paper](https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/qj.2246) as `Trap2(2,3,2)` with `δ_s = 1, δ_f = 0`. It is tested in unit tests and, if people want to try a more physically relevant experiment, they can run the `experiments/TestCase/isothermal_zonal_flow.jl`

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

valeriabarra commented:
bors try

valeriabarra commented:
bors try-

charleskawczynski commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -790,19 +833,115 @@ function ARK2GiraldoKellyConstantinescu(
+the following reference
+
+### References
+    @article{Ascher1997,
+      title = {Numerical analyses of Runge–Kutta implicit–explicit schemes
+               for horizontally explicit, vertically implicit solutions of
+               atmospheric models},
+      author = {S.-J. Lock and N. Wood and H. Weller},
+      volume = {140},
+      number = {682},
+      pages = {1654-1669},
+      year = {2014},
+      journal = {Quarterly Journal of the Royal Meteorological Society},
+      publisher = {{RMetS}}
+    }
```
Can we put this in `docs/bibliography/bib`?

valeriabarra replied:
Oh yes! I followed what was done in the other functions above.

valeriabarra replied:
Done in [71772a9](https://github.com/CliMA/ClimateMachine.jl/pull/1978/commits/71772a9907f033d5c89e474f70afcb8800378561)

valeriabarra commented:
bors try

thomasgibson commented on test/Numerics/ODESolvers/ode_tests_basic.jl:
```diff
@@ -118,7 +118,7 @@ Qexact = exactsolution(finaltime, q0, t0)
         for (method, order) in imex_methods
             for split_explicit_implicit in (false, true)
-                for variant in (LowStorageVariant(), NaiveVariant())
+                for variant in (NaiveVariant(),)
```
Oops, we commented out `LowStorageVarianet()` to speed up our local debugging. We need to restore this and add check somewhere to skip the case when `LowStorageVariant()` and Trap2 appear together here.

valeriabarra replied:
Okay, thanks @thomasgibson . I will add this back. Where should we check for this case and Trap2 to appear together?

thomasgibson replied:
Don't worry, I will push a fix to this --- I'm just testing a few things locally.

valeriabarra replied:
Thanks!

valeriabarra commented:
Note:

While testing this PR, I realized that in PR #1946  I did not check for backward compatibility, and unfortunately missed testing the `--degree` option that was already there. I'm pushing a fix to your branch right now.

thomasgibson commented:
bors try

valeriabarra commented:
@smarras79 do you have any feedback/comment on this PR? 
This can be tested by running `experiments/TestCase/isothermal_zonal_flow.jl`

valeriabarra commented:
bors try

jkozdon commented on src/Numerics/ODESolvers/BackwardEulerSolvers.jl:
```diff
@@ -67,6 +67,19 @@ solver.
+where `Q` is the state vector, `Qhat` is the right-hand side,
+and `f` is the discretized implicit operator.
+"""
+get_implicit_operator_coefficient(::AbstractBackwardEulerSolver) = nothing
```
```suggestion
function get_implicit_operator_coefficient end
```
I think that this is the what we want so an error is thrown if the method is used but not defined for the concrete type

valeriabarra replied:
Thanks!!

jkozdon commented on test/Numerics/ODESolvers/ode_tests_common.jl:
```diff
@@ -21,13 +21,29 @@ const explicit_methods = (
     (ARK437L2SA1KennedyCarpenter, 4),
     (ARK548L2SA2KennedyCarpenter, 5),
 )
+const imex_methods_naivestorage_compatible = (
+    (ARK1ForwardBackwardEuler, 1),
+    (ARK2ImplicitExplicitMidpoint, 2),
+    (ARK2GiraldoKellyConstantinescu, 2),
+    (ARK437L2SA1KennedyCarpenter, 4),
+    (ARK548L2SA2KennedyCarpenter, 5),
+    # Some methods can only be used with the `NaiveVariant` storage
+    # scheme since, in general, ARK methods can have different time-scaling/rhs-scaling
+    # coefficients (C/B vectors in the Butcher tables). For future reference,
+    # any other ARK-type methods that have more general Butcher tables
+    # (but with same number of stages) should be tested here:
+    (Trap2LockWoodWeller, 2),
+)
```
Not required, but we could splat here:

```suggestion
const imex_methods_naivestorage_compatible = (
    imex_methods_lowstorage_compatible...,
    # Some methods can only be used with the `NaiveVariant` storage
    # scheme since, in general, ARK methods can have different time-scaling/rhs-scaling
    # coefficients (C/B vectors in the Butcher tables). For future reference,
    # any other ARK-type methods that have more general Butcher tables
    # (but with same number of stages) should be tested here:
    (Trap2LockWoodWeller, 2),
)
```

jkozdon commented on test/Numerics/ODESolvers/ode_tests_convergence.jl:
```diff
@@ -669,7 +707,7 @@ const ArrayType = ClimateMachine.array_type()
 
             finaltime = 1
-            dts = [2.0^(-k) for k in 6:7]
+            dts = [2.0^(-k) for k in 6:8]
```
Why the increase?

thomasgibson replied:
`Trap2` needed an additional nudge to reach the `atol` criteria in its convergence check

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -82,10 +83,8 @@ mutable struct AdditiveRungeKutta{
-    "backward Euler solvers, pre-factorized, for MIS"
-    besolvers!::Any
+    "a dictionary of backward Euler solvers"
+    implicit_solvers::Any
```
Do we really want an `Any` here? (also saying `Any` isn't needed since its implied)

charleskawczynski replied:
I think the formatter might apply `Any`, but I agree that more concrete is better here.

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -139,49 +144,64 @@ mutable struct AdditiveRungeKutta{
+
+        # NOTE: this is only for composing an ARK method with
+        # a MIS timestepper
+        # TODO: Clean this up
```
should this `TODO` be handled?

valeriabarra replied:
Sorry, yes, we already took care of it. Soon to be pushed together with other great suggestions from you :)

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -139,49 +144,64 @@ mutable struct AdditiveRungeKutta{
-        for is in 2:Nstages
-            @assert RKA_implicit[is, is] ≈ RKA_implicit[2, 2]
-        end
+        # for is in 2:Nstages
+        #     @assert RKA_implicit[is, is] ≈ RKA_implicit[2, 2]
+        # end
```
Remove dead code? Or can we run the check is the right variant is set? (e.g., does `LowStorageVariant` imply `SDIRK`?)

jkozdon replied:
Could also check that if `LowStorageVariant ` that `RKB_implicit == RKB_explicit` and `RKC_implicit == RKC_explicit`

valeriabarra replied:
yes, we removed the dead code. Will push soon.

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -139,49 +144,64 @@ mutable struct AdditiveRungeKutta{
+        # Remove all zero entries from `rk_diag`
+        # so we build all unique implicit solvers (parameterized by the
+        # corresponding RK coefficient)
+        filter!(c -> c !== 0, rk_diag)
```
```suggestion
        filter!(c -> c != 0, rk_diag)
```

You want this to be `!=` since:

```julia
julia> 0.0 !== 0
true

julia> 0.0 != 0
false
```

valeriabarra replied:
Oh my goodness!! This was a nasty bug!! Thanks @jkozdon !!

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -790,19 +846,116 @@ function ARK2GiraldoKellyConstantinescu(
+                    split_explicit_implicit, variant)
+
+This function returns an [`AdditiveRungeKutta`](@ref) time stepping object,
+    see the documentation of [`AdditiveRungeKutta`](@ref) for arguments definitions.
+    This time stepping object is intended to be passed to the `solve!` command.
```
```suggestion
see the documentation of [`AdditiveRungeKutta`](@ref) for arguments definitions.
This time stepping object is intended to be passed to the `solve!` command.
```

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -790,19 +846,116 @@ function ARK2GiraldoKellyConstantinescu(
+    @assert dt !== nothing
+    # In this scheme B and C vectors do not coincide,
+    # hence we can't use the LowStorageVariant optimization
+    @assert variant !== LowStorageVariant()
```
```suggestion
    @assert variant isa NaiveVariant
```

jkozdon reviewed (approved):
Overall this seems fine to me.

valeriabarra commented:
Hey @jkozdon , thanks so much for your thorough review! We think we addressed all comments. If you want to give another quick look/review, it would be great! Thanks!

valeriabarra commented:
bors try

valeriabarra commented:
ooops! We thought we were done and tested this locally, but obviously we missed/mangled something! Will work on it more

valeriabarra commented:
bors try

valeriabarra commented:
bors try

smarras79 commented:
> @smarras79 do you have any feedback/comment on this PR?
> This can be tested by running `experiments/TestCase/isothermal_zonal_flow.jl`

No comments on this PR from my end. Thanks for putting it together and making it so clean.

thomasgibson commented:
bors try

valeriabarra commented:
Thanks @thomasgibson for the fix!! I can go ahead and reduce the number of commits in this PR and keep authorship and if @jkozdon wants to give a final look/review we'd appreciate it.

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -82,10 +84,10 @@ mutable struct AdditiveRungeKutta{
-    besolvers!::Any
+    "a dictionary of backward Euler solvers"
+    implicit_solvers::IST
+    "An integer which is updated to determine the appropriate cached implicit solver (used in substepping)"
```
Wrap long comment
```suggestion
    "An integer which is updated to determine the appropriate cached implicit
    solver (used in substepping)"
```

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -137,51 +145,75 @@ mutable struct AdditiveRungeKutta{
+            @assert RKC_explicit == RKC_implicit
+            # rk_diag here has been filtered of all non-unique and zero values.
+            # So [0, c, ... c ] filters to [c]
+            if length(rk_diag) != 1
+                error("The implicit Butcher table must have SDIRK form when using LowStorageVariant.")
+            end
```
I'd be ok with this being an `@assert`
```suggestion
            @assert length(rk_diag) == 1
```
Since all the others checks above are `@asserts`

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -137,51 +145,75 @@ mutable struct AdditiveRungeKutta{
+                    ),
+                    nsteps,
+                )
+                implicit_solvers[rk_coeff] = solvers
```
Could add:
```julia
@assert(all(isa.(solvers, AbstractBackwardEulerSolver)))
```
to match above check.

jkozdon commented on src/Numerics/ODESolvers/AdditiveRungeKuttaMethod.jl:
```diff
@@ -790,19 +865,116 @@ function ARK2GiraldoKellyConstantinescu(
+    RT = real(T)
+
+    #! format: off
+    RKA_explicit = [
+        RT(0)    RT(0)    RT(0)    RT(0)
+        RT(δ_s)  RT(0)    RT(0)    RT(0)
+        RT(1/2)  RT(1/2)  RT(0)    RT(0)
+        RT(1/2)  RT(0)    RT(1/2)  RT(0)
+    ]
+
+    RKA_implicit = [
+        RT(0)                  RT(0)                  RT(0)    RT(0)
+        RT(δ_f * (1 - α) / 2)  RT(δ_f * (1 + α) / 2)  RT(0)    RT(0)
+        RT(1/2)                RT(0)                  RT(1/2)  RT(0)
+        RT(1/2)                RT(0)                  RT(0)    RT(1/2)
+    ]
```
Super pedantic, but can we put spaces between the numbers and division (to match the formatter?)

Also, you could use rational `1 // 2` before casting to floating point type.

valeriabarra replied:
Thanks! I will add the space, but I'd rather follow the convention used in the other ARK methods, where I didn't see using the rational division ` // `. I would keep a single `/` for consistency.

jkozdon reviewed (approved):
Couple nitpicky suggestions, but nothing major (feel free to skip them).

But looks good. Thanks!

valeriabarra commented:
Thanks @jkozdon for your final review. 

@thomasgibson , I took care of those suggested small fixes. You can `bors r+` at your convenience!

valeriabarra commented:
bors r+
